### PR TITLE
Health

### PR DESCRIPTION
**Pull request by Captain4LK**
* Start health, make asteroids destruction health based
* Add health to enemy ships
* Add health to player
* Adapt new tick() behaviour properly in ai
* Add damage variable to entity
* Fix enemy bullet damage being set to 1 instead of ent->damage
* Remove redundant comment

### CODE REVIEW

**Review by cedric-h — COMMENTED** [review]
It looks good, this is an important step towards a playable game! I'm worried the `remove_ent` call might be a bug. Aside from that, I'm a little bit concerned -- perhaps prematurely -- about the code duplication and the rapidly growing number of responsibilities that `EntProp_AI` has. Those minor qualms aside, though, great work.

**Comment by cedric-h** [comment]
I'm ashamed to admit I may have rubber stamped the original AI PR, but hopefully I can make up for it by giving a more thorough review of some of the code that was introduced in that PR now.

# Superficial Concerns
> `    _ai_set_state(ent, _ai_entinfo[ent->ai.type].state_death);`

Every call to `ai_set_state` follows this pattern except for the one that passes `ent->ai.state->next`. To narrowly hone in on just this issue, one could alleviate this duplication by changing
```c
typedef struct {
  AI_statenum state_idle;
  AI_statenum state_move;
  AI_statenum state_attack;
  AI_statenum state_death;
} AI_info;
```
to
```c
typedef enum {
  AI_Info_State_Idle;
  AI_Info_State_Move;
  AI_Info_State_Attack;
  AI_Info_State_Death;
  AI_Info_COUNT;
} AI_Info;
```
and use it like so:
```c
typedef struct { AI_statenum info[AI_info_COUNT]; } _ai_EntFo;

static const _ai_EntFo _ai_entinfo[AI_TYPE_MAX] = {
  //AI_TYPE_DSHIP
  {
    [AI_Info_State_Idle  ] = AI_STATE_SHIP_IDLE,
    [AI_Info_State_Move  ] = AI_STATE_SHIP_MOVE,
    [AI_Info_State_Attack] = AI_STATE_SHIP_ATTACK0,
    [AI_Info_State_Death ] = AI_STATE_SHIP_DEATH,
  },
  // ...
```
which would allow you to make a wrapper around `ai_set_state` which finds the appropriate `nstate` for the Ent's type, while minimizing duplication:
```c
static void _ai_set_entfo_state(void *param0, AI_info info_state) {
  Ent *ent = (Ent *)param0;
  ai_set_state(ent, _ai_entinfo[ent->ai.type][info_state]);
}
```
so now our original call,
> `    _ai_set_state(ent, _ai_entinfo[ent->ai.type].state_death);`

can become simply
`ai_set_entfo_state(ent, AI_Info_State_Death);`

# Missing the Forest For the Trees
However, I couldn't in good conscience merge such code, not without ignoring the bigger picture. Removing a few lines of gnarly syntax on almost all invocations of `ai_set_state` is nice, but it's not going to let you express anything that you couldn't before; it's essentially just syntactic sugar. The whole problem `AI_Info_State` addresses is that `AI` is doing such drastically different things for different kinds of entities, that core functions need to be swapped out to facilitate that. Your original `AI_info` struct was a clever way to go about that, but "clever" is unfortunately often at odds with "elegant."

# Falling into place like Dominoes
In my opinion, the "elegant" solution here would be to separate the loot behavior from the AI behavior. Then, asteroids would have no need for AI. That could be implemented separately in another `EntProp` that has its own small, clearly defined set of responsibilities. All of the "clever" but ultimately gnarly `AI_Info` code could be deleted, as you wouldn't need to swap out the core behavior of the AI property because it only has one core behavior. The number of lines of code in the program could go down drastically.

File: collision.h [hunk 1]
```diff
@@ -67,31 +67,9 @@ static void collision(Ent *ac) {
...
-        }
+    ent->health-=ac->damage;
 
-        ne->pos = add2_f(old.pos, old.collider.size/2.0f * sign);
-        ne->vel = mul2_f(old.vel, sign);
-
-        if (has_ent_prop(ne, EntProp_PassiveRotate))
-          ne->passive_rotate_axis = rand3();
-      }
+    remove_ent(ac);
```
[the remark this answers is not in the document]
Comment (Captain4LK) [reply collision.h:72]: This removes the projectile, not the entity taking damage

[the remark this answers is not in the document]
Comment (cedric-h) [reply collision.h:72]: haha, can't believe I missed that. alright, great!

**Pull request merged**